File: source_code/general03.cpp
// general03.cpp
// friend class

#include <iostream>
using namespace std;

// class declaration
class Pontikaros;

class Pithari {
private:
    int num_of_fruits;
    int num_of_biscuits;
    int num_of_pies;

public:
    // constructor to initialize numB to 1
    Pithari(int a, int b, int c) : num_of_fruits(a), num_of_biscuits(b), num_of_pies(c) {}

    // get values
    void show_content() {
        cout << num_of_fruits << " fruits\n" << num_of_biscuits << " biscuits\n" << num_of_pies << " pies\n";
    }

    // member function to add numA
    friend class Pontikaros;
};

class Pontikaros {
private:
    bool hungry;
    int weight; // in kilos

public:
    // constructor
    Pontikaros() : hungry(true), weight(2) {}

    void eat_from(Pithari* p) {
        if (hungry) {
            p->num_of_fruits -= 2;   // eat 2 fruits
            p->num_of_biscuits -= 3; // eat 2 biscuits
            p->num_of_pies -= 1;     // eat 1 pie

            hungry = false;
            weight += 3;
        }
    }
};

int main() {
    Pithari pithos(5, 5, 2);
    Pontikaros Mitsos;
    
    cout << "Pithari before:\n";
    pithos.show_content();

    Mitsos.eat_from(&pithos);
    
    cout << "\nPithari after:\n";
    pithos.show_content();

    return 0;
}

File: source_code/general03.py
# general03.py
#

class Pithari():
    def __init__(self, a, b, c):
        self._num_of_fruits = a
        self._num_of_biscuits = b
        self._num_of_pies = c

    def show_content(self):
        print(self._num_of_fruits, " fruits\n", self._num_of_biscuits, " biscuits\n", self._num_of_pies, " pies\n")

class Pontikaras():
    def __init__(self):
        self._hungry = True
        self._weight = 2    # kilos

# create objects
obj00 = Pithari(50, 86, 12)

obj01 = Pontikaras()
obj01._name = "Mitsaras"


# create a method
def eat_from(self, Pith):
    Pith._num_of_fruits -= 2
    Pith._num_of_biscuits -= 3
    Pith._num_of_pies -= 1

    self._hungry = False
    self._weight += 3

    print("\n", self._name, " just ate from pithos\n")

# add/bound the method to the object
obj01.eat_from = eat_from.__get__(obj01)

print("Pithari before:\n")
obj00.show_content();

# and it works!?!?!???
obj01.eat_from(obj00)

print("Pithari after:\n")
obj00.show_content();
